### ISSUE TITLE
Please revert the deprecation to declare additional feature variants on the `main` source set

### ISSUE THREAD

**Issue by Vampire** [gradle/gradle#29455]
### Current Behavior

The feature is deprecated.

### Expected Behavior

The feature should continue to work as it was used to.

### Context (optional)

One of the use-cases for feature variants are optional dependencies, as also described on https://blog.gradle.org/optional-dependencies.

One specialization of this use-case is to automatically add dependencies depending on some attribute like target JVM version.

For these use-cases, there is no custom code for the feature variant and it should indeed use the same code jar.
This is not - or at least not too obvious anymore - possible if this deprecated feature is actually removed.

If there will be a way and the deprecation and removal stays, at least that blog post should be updated to show the new idiomatic way as the blog post is still quite relevant.

### Steps to Reproduce

This build publishes two feature variants with the same code jar that are automatically selected depending on target jvm version and just add a runtime dependency for Java 9+.

```kts
plugins {
    `java-library`
    `maven-publish`
}

group = "showcase"
version = "0.1.0-SNAPSHOT"

publishing {
    val library by publications.registering(MavenPublication::class) {
        from(components["java"])
    }
}

java {
    targetCompatibility = JavaVersion.VERSION_1_8
    registerFeature("java9") {
        usingSourceSet(sourceSets.main.get())
        capability("$group", name, "$version")
    }
}
val java9ApiElements by configurations.existing {
    attributes {
        attribute(TargetJvmVersion.TARGET_JVM_VERSION_ATTRIBUTE, 9)
    }
}
val java9RuntimeElements by configurations.existing {
    attributes {
        attribute(TargetJvmVersion.TARGET_JVM_VERSION_ATTRIBUTE, 9)
    }
}
val java9RuntimeOnly by configurations.existing
dependencies {
    java9RuntimeOnly("javax.cache:cache-api:1.1.1")
}
```

### Gradle version

8.8

**Comment by Vampire** [comment]
I think this should be the work-around if the deprecation is not going to be reverted:
```kts
plugins {
    `java-library`
    `maven-publish`
}

group = "showcase"
version = "0.1.0-SNAPSHOT"

publishing {
    val library by publications.registering(MavenPublication::class) {
        from(components["java"])
    }
}

val java9 by sourceSets.creating
java {
    targetCompatibility = JavaVersion.VERSION_1_8
    registerFeature("java9") {
        usingSourceSet(java9)
        capability("$group", name, "$version")
    }
}
val java9ApiElements by configurations.existing {
    attributes {
        attribute(TargetJvmVersion.TARGET_JVM_VERSION_ATTRIBUTE, 9)
    }
}
val java9RuntimeElements by configurations.existing {
    attributes {
        attribute(TargetJvmVersion.TARGET_JVM_VERSION_ATTRIBUTE, 9)
    }
}
val java9RuntimeOnly by configurations.existing
dependencies {
    java9RuntimeOnly("javax.cache:cache-api:1.1.1")
}

val apiElements by configurations.existing
val java9ApiElements by configurations.existing {
    outgoing.artifacts.clear()
    outgoing.variants.clear()
    outgoing.variants.addAllLater(apiElements.map { it.outgoing.variants })
    extendsFrom(apiElements.get())
}
val runtimeElements by configurations.existing
val java9RuntimeElements by configurations.existing {
    outgoing.artifacts.clear()
    outgoing.variants.clear()
    outgoing.variants.addAllLater(runtimeElements.map { it.outgoing.variants })
    extendsFrom(runtimeElements.get())
}
```

**Comment by ljacomet** [comment]
This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by big-guy** [comment]
I can see why this is useful, but we think it's not safe because tests run against the feature's dependencies, but the main source set variants do not have those dependencies. This means code tested with these dependencies can then run without those dependencies. This gives you a false sense of correctness. 

Additionally, if the project has multiple test suites (different test source sets), you don't get the optional dependencies or you need to wire the configurations manually or you need to know how to request the right variant of the current project via dependency management. 

We want to support multi-version targets in the JVM plugins, but this isn't the mechanism we want to use.  This deprecation unblocks us making changes with how source sets interact with variants.

**Comment by Vampire** [comment]
@big-guy
I understand what you say here and in the cousin-issue regarding multi-Java-target-version.
But that was just one common example. Another is the also mentioned optional dependencies that the blog post also describes.
People are still reading and following and recommending that blog post, but if you follow its instructions now, you get a deprecation warning and after the deprecation cycle it will fail.

May I kindly ask how this should be solved instead then and maybe the blog post to be updated,
or a significant warning be added and maybe a followup post with the new idiomatic way?
Should instead the feature variant depend on the main variant or how is it to be done?

**Comment by twbecker** [comment]
@big-guy I just ran across this as well and the documentation around the deprecation is seriously deficient. Adding another source set changes the entire source and artifact layout of the project. In contrast to @Vampire use-case, another common use for variants on the `main` source set is - as described in the [user manual](https://docs.gradle.org/current/userguide/feature_variants.html) - 

    a (better) substitute for Maven optional dependencies

As far as I can see this use-case is completely broken now. Because the classes that require the optional dependency are now in an entirely separate jar - one that will not be pulled in by existing consumers.

IMO there needs to continue to be a way to generate a single jar that simply declares certain dependencies as optional in the metadata.

**Comment by mindhaq** [comment]
I second that this deprecation should be reverted.

I am using this to publish an (internal) utility library with some dependencies, which can be added by specifying a `requireCapability` and must not be added for all consumers of that library.

The [deprecation documentation](https://docs.gradle.org/8.13/userguide/upgrading_version_8.html#deprecate_register_feature_main_source_set) is not helping me at all for this case.

**Comment by jjohannes** [comment]
Most compact solution I found:
```
java.registerFeature("signing") {
    // register additional source set as workaround for deprecation:
    // https://docs.gradle.org/8.13/userguide/upgrading_version_8.html#deprecate_register_feature_main_source_set
    usingSourceSet(sourceSets.create("dummy"))
    // set to default capability
    capability(project.group.toString(), project.name, project.version.toString())
}

configurations.matching { it.name in listOf("apiElements", "runtimeElements") }.configureEach {
    // use compiled code from "main" source set instead of (non-existing) "dummy" source set
    outgoing.artifacts.clear()
    outgoing.artifact(tasks.jar)
    outgoing.variants.named("classes") {
        artifacts.clear()
        artifact(sourceSets.main.get().java.classesDirectory)
    }
}
```

Alternatively, one could set this up via more "low level" capability APIs but that's quite fragile as you need to copy quite some code otherwise done by the Java plugins. Such setting the correct attributes on the Configurations.


My 2ct: I think deprecation this without a proper replacement is wrong as there is, in my understanding, a clear use case that is [described in the official blog post on the topic](https://blog.gradle.org/optional-dependencies) that is now no longer supported directly.
In particular in open source development. it is not always an option to split classes into two Jar files as it may break consumers that do not support Gradle Metadata (such as Maven).
And there are cases where you have no additional classes for a second source set. For example, if you just want to pull in different implementations at runtime that are then loaded via the service loader.

**Comment by Vampire** [comment]
Your "Most compact solution" is quite bad though. :-(
- You do not modify the signing configurations, but the main configurations.
- You miss to set the artifact type on the classes variant.
- You miss to handle the resources variant at all.

So to fix your "Most compact solution" it would be
```kotlin
java.registerFeature("signing") {
    // register additional source set as workaround for deprecation:
    // https://docs.gradle.org/8.13/userguide/upgrading_version_8.html#deprecate_register_feature_main_source_set
    usingSourceSet(sourceSets.create("signing"))
}

configurations.matching { it.name in listOf("signingApiElements", "signingRuntimeElements") }.configureEach {
    // use compiled code from "main" source set instead of (non-existing) "signing" source set
    outgoing.artifacts.clear()
    outgoing.artifact(tasks.jar)
    outgoing.variants.named("classes") {
        artifacts.clear()
        artifact(sourceSets.main.get().java.classesDirectory) {
            type = JVM_CLASS_DIRECTORY
        }
    }
    outgoing.variants.named { it == "resources" }.configureEach {
        artifacts.clear()
        artifact(sourceSets.main.get().output.resourcesDir!!) {
            builtBy(sourceSets.main.get().processResourcesTaskName)
            type = JVM_RESOURCES_DIRECTORY
        }
    }
}
```

An even more compact way that needs way less knowledge about the internals that achieves the same would be
```kotlin
java.registerFeature("signing") {
    // register additional source set as workaround for deprecation:
    // https://docs.gradle.org/8.13/userguide/upgrading_version_8.html#deprecate_register_feature_main_source_set
    usingSourceSet(sourceSets.create("signing"))
}

configurations.matching { it.name in listOf("signingApiElements", "signingRuntimeElements") }.configureEach {
    val main = configurations.named(name.substringAfter("signing").replaceFirstChar { it.lowercase() })
    outgoing.artifacts.clear()
    // the next is not necessary with `extendsFrom(main.get())`
    // as also outgoing artifacts are inherited and thus would be there twice
    outgoing.artifacts(main.map { it.outgoing.artifacts })
    outgoing.variants.clear()
    outgoing.variants.addAllLater(main.map { it.outgoing.variants })
}
```

Which basically is exactly what I wrote above with my Java 9+ example, just that it does not add the variants and uses a bit different syntax with property delegates instead of `matching`.
And of course does need to set capability and attributes and extend the main configuration to get the same dependencies, but all that would be necessary here too. :-)

**Comment by 3ll3d00d** [comment]
> We want to support multi-version targets in the JVM plugins, but this isn't the mechanism we want to use. This deprecation unblocks us making changes with how source sets interact with variants.

deprecation is one thing, removal a year later without actually providing a replacement is a rather hostile move which adds to the feeling that gradle plays too fast and loose with its api 

the rest of your post is essentially "gradle knows better than the user"

the situation here is really v poor

**Comment by chadlwilson** [comment]
With Gradle 9 now out, are we still in a position of deprecation-without-reasonable-replacement?

I just spent a whole lot of time refactoring a hundred-project setup to consume some variants like the blog example to manage optional dependencies better - just to find out it's deprecated (I have hundreds to fix, and it's non trivial to get through them and check more aren't appearing accidentally). This is super-frustrating. I hope there are idiomatic alternatives documented somewhere by now?

The suggested workarounds above don't seem to work for me as I'd eitehr need to split all the source code up further as the main source set cannot be compiled without the "optionals" or litter dependencies with `compileOnly` like the old code used to do. I'm trying to incrementally make the risk of optional dependencies lower and more clearly modeled - not "boil the ocean" with perfect modelling - which the deprecation here unfortunately seems to be requiring.

Earlier the project was using dodgy `compileOnly`s and required other consuming projects/libraries to add their own dependency if they used certain classes which was completely opaque and error-prone. The "now deprecated" mechanism of using the main sourceSet with a feature was an incremental improvement, but seems now I'd need to add a chunk of workaround boilerplate as well as the `compileOnly` and `testCompileOnly` dependencies - or completely restructure code?

**Comment by 3ll3d00d** [comment]
> I hope there are idiomatic alternatives documented somewhere by now?

not that I could find

I have a similar setup (with a similarly large no of projects involved) & decided to use a version catalogue to capture those optional dependencies as bundles, from a user (developer of consuming project) point of view this seems broadly equivalent (previously they had to add 1 or more capabilities, now they have to add 1 or more dependency bundles).

however version catalogues themselves seem incomplete and have various wrinkles around them, both in gradle and from intellij (not sure about other IDEs). The main ones

* https://youtrack.jetbrains.com/issue/IDEA-377528 from IDE point of view
* that composite builds don't support substitution of published version catalogues (mainly an issue for the catalogue author rather than end developer) as per #18847 
* version catalogues don't support classifiers and the syntax for adding that in a build is extremely non obvious

**Comment by Vampire** [comment]
> With Gradle 9 now out, are we still in a position of deprecation-without-reasonable-replacement?

Seems so.
At least the removal was delayed, it is still deprecated in 9 and now states it will be removed in 10.

The replacements probably remain to be https://github.com/gradle/gradle/issues/29455#issuecomment-2154629588 or https://github.com/gradle/gradle/issues/29455#issuecomment-2852827469, it would just nice if the blog post got updated, or a new blog post added and at least a warning that the recommendations are deprecated in the existing one added.

> The suggested workarounds above don't seem to work for me as I'd eitehr need to split all the source code up further as the main source set cannot be compiled without the "optionals" or litter dependencies with compileOnly like the old code used to do.

If you actually have feature-specific code, it would even without the deprecation most probably be cleaner to split this into separate source sets and thus artifacts. But to avoid that, I guess (did not try yet) you can make `compileClasspath` or `compileOnly` extend from `yourFeatureCompileClasspath` so that the dependencies you declare on `yourFeature` also land in the compile classpath of the main source set without duplication or leaking into the main variant's metadata.

**Comment by chadlwilson** [comment]
Yes, it'd be cleaner to split. But then source in one "extension" needs the other so that makes independent source sets inconvenient and I may as well factor it into entirely different projects rather than rely on fancy Gradle features/concepts which (apparently) can be deprecated after a few years.

This is a large 18 year old OSS distributable software project I have inherited and choose to maintain for the community on my own. Spending weeks rejigging build tooling is not a valuable use of time, except where it serves some other personal learning purpose.

**Comment by Vampire** [comment]
Definitely. :-)
But as I said, I _think_ it should be sufficient to set up proper `extendsFrom`.

**Comment by chadlwilson** [comment]
Yes, thanks @Vampire - I'll have another go and see how 'ugly' it gets.

**Comment by Vampire** [comment]
Let us know how it went :-)

**Comment by chadlwilson** [comment]
Well...... it might be a bad idea, but this seems to have worked (still using Groovy because... hundreds of projects):

in `buildSrc/src/main/groovy` for now for easy re-use while hiding gross boiler-plate away somewhere I won't see it :-)

```groovy

import org.gradle.api.Project
import org.gradle.api.artifacts.Configuration

import static org.apache.commons.lang3.StringUtils.substringAfter
import static org.apache.commons.lang3.StringUtils.uncapitalize

class CapabilitySupport {
  /**
   * Workaround to deprecated functionality to use the main sourceSet with feature variants from https://github.com/gradle/gradle/issues/29455
   * See https://docs.gradle.org/8.13/userguide/upgrading_version_8.html#deprecate_register_feature_main_source_set
   *
   * Configures each feature's configurations to include the same as for the `main` sourceSet, and allows the main sourceSet
   * to be compiled using the feature specific dependencies.
   */
  static def extendMain(Project project, String... features) {
    project.with {
      features.each { feature ->
        // Make feature configurations include the the same artifacts and variants as main
        configurations.matching { it.name == "${feature}ApiElements" || it.name == "${feature}RuntimeElements" }.configureEach { c ->
          def main = configurations.named(uncapitalize(substringAfter(name, feature)))
          c.outgoing.artifacts.clear()
          c.outgoing.artifacts(main.map { it.outgoing.artifacts })
          c.outgoing.variants.clear()
          c.outgoing.variants.addAllLater(main.map { it.outgoing.variants })
        }

        // Make the "main" classpath configurations able to consume the feature's dependencies
        ['compileClasspath', 'testCompileClasspath'].each { c ->
          configurations.named(c).configure { mc ->
            mc.extendsFrom(configurations."${feature}CompileClasspath" as Configuration)
          }
        }
        ['testRuntimeClasspath'].each { c ->
          configurations.named(c).configure { mc ->
            mc.extendsFrom(configurations."${feature}RuntimeClasspath" as Configuration)
          }
        }
      }
    }
  }
}
```

In various projects' `build.gradle`

```groovy
java {
  // For use of the LogConfigurator or LogHelper
  registerFeature('logging') {
    usingSourceSet(sourceSets.create('logging'))
  }

  // For use of the GoAgentServerClientBuilder
  registerFeature('agentClient') {
    usingSourceSet(sourceSets.create('agentClient'))
  }
}

CapabilitySupport.extendMain(project, 'logging', 'agentClient')

dependencies {
  api project.deps.slf4jApi
  loggingApi project.deps.logback

  agentClientApi project.deps.logback
  agentClientImplementation project.deps.bouncyCastle
  agentClientImplementation project.deps.bouncyCastlePkix
  agentClientImplementation project.deps.apacheHttpComponents
  agentClientRuntimeOnly project.deps.slf4jJcl // Allow commons-logging replacement

  // etc
}
```

with feature variants being consumed by other projects with things like

```gradle
dependencies {
  api(project(':base')) { capabilities { requireFeature('logging') } }
}
```
